### PR TITLE
Create 2020-12-17-solidity-v0.8.0-release-announcement.md

### PR DESCRIPTION
**Pull request by franzihei**
still contains some of the old text from the 0.7.0 announcement, which needs to be adjusted

### CODE REVIEW

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+layout: post
+published: true
+title: 'Solidity 0.8.0 Release Announcement'
+date: '2020-12-17'
+author: Solidity Team
+category: Releases
+---
+
+[Solidity 0.8.0](https://github.com/ethereum/solidity/releases/tag/v0.8.0) is a breaking release of the Solidity compiler and language.
+
```
Comment (axic) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:11]: Should we refer to the "preview 0.8 release" and mark it superseded, and also perhaps mention if we did a major change to the features contained within?

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+This release does not include many features but rather changes that require a
+backwards-incompatible adjustment in syntax or semantics. For a detailed explanation,
+please see the [documentation](https://docs.soliditylang.org/en/latest/080-breaking-changes.html).
+
+The change that will affect most users is that arithmetic operations are now checked by default,
+which means that overflow and underflow will cause a revert. This feature can be disabled locally.
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:17]: The first question that popped into my mind when I read this was "how" so I think I'd add this:

```suggestion
which means that overflow and underflow will cause a revert. This feature can be disabled locally by using an ``unchecked`` block.
```

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+A second change that is very visible is that the ABI coder v2 is activated by default. You can
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:19]:
```suggestion
The second change that is very visible is that the ABI coder v2 is activated by default. You can
```

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+Furthermore, internal errors like division by zero, assertion failure and others
+do not use the invalid opcode anymore, but use ``revert`` with a special error message,
+in order to not waste gas in such situations.
+
+Another important change is that we restricted the possibilities of explicit conversions
+to avoid ambiguities. All conversions that were possible before are still possible, but
+you might have to perform two conversions to get there - which does not affect the generated code, though.
+
+These are only some of the more important breaking changes in 0.8.0., please read the changelog below for the full list!
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:31]:
```suggestion
These are only some of the more important breaking changes in 0.8.0, please read the changelog below for the full list!
```

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+### Checked Arithmetic
+
+The "Checked Arithmetic" feature of Solidity 0.8.x consists of three sub-features:
+- Revert on assertion failures and similar conditions instead of using the invalid opcode.
+- Actual checked arithmetic.
+- ``unchecked`` blocks.
+
+#### Revert on assertion failures and similar conditions instead of using the invalid opcode
+
+Prevously, internal errors like division by zero, failing assertions, array access out of bounds,
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:42]:
```suggestion
Previously, internal errors like division by zero, failing assertions, array access out of bounds,
```

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+etc., would result in an invalid opcode being executed. The idea was to distinguish
+these more critical errors from non-critical errors that lie outside of the domain
+of the Smart Contract programmer like invalid input, not enough balance for a token transfer,
+and so on. These non-critical errors use the ``revert`` opcode and optionally add an error description.
+
+The problem with the invalid opcode is that, in contrast to the ``revert`` opcode, it consumes all gas available.
+This makes it very expensive and you should try to avoid it at all cost.
+
+We wanted to consider arithmetic overflow as a critical error, but did not want to cause it to consume all gas.
+As a middle ground, we chose to use the ``revert`` opcode, but provide a different error data
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:52]:
```suggestion
As a middle ground, we chose to use the ``revert`` opcode, but provide different error data
```

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+so that static (and dynamic) analysis tools can easily distinguish them.
+
+More specifically:
+ - Non-critical errors will revert either with empty error data or with error data that
+   corresponds to the ABI-encoding of a function call to a function with the signature ``Error(string)``.
+ - Critical errors revert with error data that corresponds to the ABI-encoding of a function call
+   to a function with the signature ``Panic(uint256)``.
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:59]: I see that github renders the list correctly in this case but indenting top-level list items sometimes results in a verbatim block instead so it's better to avoid it.

```suggestion
- Non-critical errors will revert either with empty error data or with error data that
    corresponds to the ABI-encoding of a function call to a function with the signature ``Error(string)``.
- Critical errors revert with error data that corresponds to the ABI-encoding of a function call
    to a function with the signature ``Panic(uint256)``.
```

Some lists below also have this extra indent.

Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:59]:
```suggestion
   corresponds to the ABI-encoding of a function call with the signature ``Error(string)``.
 - Critical errors revert with error data that corresponds to the ABI-encoding of a function call
   with the signature ``Panic(uint256)``.
```

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+To distinguish the two situations, you can take a look at the first four bytes of the return data in case
+of a failure. If it is ``0x4e487b71``, then it is a panic, if it is ``0x08c379a0``
+or if the message is empty, it is a "regular" error.
+
+Panics use specific error codes to distinguish certain situations in which the panic is triggered.
+This is the current list of error codes:
+
+1. 0x01: If you call ``assert`` with an argument that evaluates to false.
+2. 0x11: If an arithmetic operation results in underflow or overflow outside of an ``unchecked { ... }`` block.
+3. 0x12: If you divide or modulo by zero (e.g. ``5 / 0`` or ``23 % 0``).
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:72]:
```suggestion
3. 0x12: If you divide or divide modulo by zero (e.g. ``5 / 0`` or ``23 % 0``).
```

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+4. 0x21: If you convert a value that is too big or negative into an enum type.
+5. 0x31: If you call ``.pop()`` on an empty array.
+6. 0x32: If you access an array, ``bytesN`` or an array slice at an out-of-bounds or negative index (i.e. ``x[i]`` where ``i >= x.length`` or ``i < 0``).
+7. 0x41: If you allocate too much memory or create an array that is too large.
+8. 0x51: If you call a zero-initialized variable of internal function type.
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:77]: The list is not the same as [the one we have in the docs](https://docs.soliditylang.org/en/breaking/control-structures.html#panic-via-assert-and-error-via-require).

`0x22` is missing:
```markdown
5. 0x22: If you access a storage byte array that is incorrectly encoded.
```

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+
+ - ``uint16(int16(int8(-1)))`` results in ``0xffff``, while
+ - ``uint16(uint8(int8(-1)))`` is ``0xff``.
+
+There are more such changes in relation to payable addresses, literals, enums and contracts. For the full list
+of changes, please see the [documentation](https://docs.soliditylang.org/en/latest/080-breaking-changes.html).
+
+### General Remarks
+
+Since we usually do not backport bugfixes, it is recommended to upgrade all code to be compatible with Solidity v.0.8.0.
```
Comment (axic) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:213]:
```suggestion
Since we usually do not backport bugfixes, it is recommended to upgrade all code to be compatible with Solidity v0.8.0.
```

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+We did not introduce checks for explicit type conversions from larger to smaller
+types, because we think that such checks might be unexpected. We would love to get
+your input on that matter!
+
+The following operations now have checks that they did not have before:
+ - addition (``+``), subtraction (``-``), multiplication (``*``)
+ - increment and decrement (``++`` / ``--``)
+ - unary negation (``-``)
+ - exponentiation (``**``)
+ - division (see below) (``/``)
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:114]: Should we mention `+=`, `-=`, `*=`, `/=`, `%=` too?

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+that it is rather expensive to do that because the EVM does not provide overflow signalling.
+Essentially you have to implement your own ``exp`` routine without using the ``exp`` opcode
+for the general case.
+
+We hope that we found a rather efficient implementation and would also appreciate your feedback about that!
+
+For many special cases, we actually implemented it using the ``exp`` opcode instead of our own implementation.
+More specifically, exponentiation operations that use a literal number as base will use the ``exp`` opcode directly.
+There are also specialized code paths for bases that are variables with small value:
+All bases up to 2, bases between 3 and 10 and bases between 11 and 306 get special treatment
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:148]:
```suggestion
all bases up to 2, bases between 3 and 10 and bases between 11 and 306 get special treatment
```

Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:148]: Between 0 and 2, right? It can't be true for all negative bases...

Also, to me this reads either as if all bases up to and including 306 get special treatment or as if 3, 10, 11 and 306 are excluded. It's supposed to say that 2 and 10 are excluded right? 2 is obvious because it can be done with bit shifts but why 10? Is there a special `exp10` opcode in the EVM for that?

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+and should result in a cheap ``exp`` opcode.
+
+#### ``unchecked`` Blocks
+
+Since checked arithmetic uses more gas, as discussed in the previous section, and
+because there are times where you really want the wrapping behaviour, for example when
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:154]:
```suggestion
because there are times when you really want the wrapping behaviour, for example when
```

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+contract C {
+    function f() public pure returns (uint) {
+        uint x = 0;
+        unchecked { x--; }
+        return x;
+    }
+}
+```
+
+An unchecked block can of course contain any number or any kind of other statements
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:171]:
```suggestion
An unchecked block can of course contain any number or any other kind of statement
```

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+2. You can only use it inside a block but not as a replacement for a block. The following code snippets, for example, are both invalid:
+
+```solidity
+function f() unchecked { uint x = 7; }
+```
+
+```solidity
+function f() pure {
+    uint x;
+    for (uint i = 0; i < 100; i ++) unchecked { x += i; }
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:190]:
```suggestion
    for (uint i = 0; i < 100; i++) unchecked { x += i; }
```

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+```
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:191]: An indent is needed here if it's meant to be a part of the second bullet point:

```suggestion
    ```solidity
    function f() unchecked { uint x = 7; }
    ```

    ```solidity
    function f() pure {
        uint x;
        for (uint i = 0; i < 100; i ++) unchecked { x += i; }
    ```
```

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+
+You have to wrap the unchecked blocks inside another block to make it work.
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:193]:
```suggestion
You have to wrap the unchecked blocks in another block to make it work.
```

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md (hunk 10 above)
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:213]:
```suggestion
Since we usually do not backport bugfixes, it is recommended to upgrade all code to be compatible with Solidity v0.8.0.
```

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md (hunk 1 above)
Comment (cameel) [reply _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:11]: Sounds like a good idea. We could mark new changes in bold or something like that in the changelog at the bottom of the post.

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md (hunk 7 above)
Comment (chriseth) [reply _posts/2020-12-17-solidity-v0.8.0-release-announcement.md]: Strictly, the function has a signature, not the call.

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md (hunk 12 above)
Comment (chriseth) [reply _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:149]: It's a full sentence, so I think capital `A` is correct.

Comment (chriseth) [reply _posts/2020-12-17-solidity-v0.8.0-release-announcement.md]: Please check again, I made some adjustments.

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md (hunk 14 above)
Comment (chriseth) [reply _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:174]: Dropped the `other`

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md (hunk 1 above)
Comment (chriseth) [reply _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:11]: I think marking entries bold could look weird.

Comment (chriseth) [reply _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:11]: I don't expect many have tried the preview release.

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md (hunk 7 above)
Comment (cameel) [reply _posts/2020-12-17-solidity-v0.8.0-release-announcement.md]: True. It does sound a bit long-winded though :)

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md (hunk 12 above)
Comment (cameel) [reply _posts/2020-12-17-solidity-v0.8.0-release-announcement.md]: TIL! https://www.grammarly.com/blog/capitalization-after-colons/

I always assumed that a colon does not end the sentence but apparently that's another part of the language where opinions vary.

Comment (cameel) [reply _posts/2020-12-17-solidity-v0.8.0-release-announcement.md]: I didn't really know whether 10 and 11 are included in the range or not so I went looking for that PR (https://github.com/ethereum/solidity/pull/9635#discussion_r488700370 in case anyone else is confused by this) and now I understand why it's stated that way. I think that it would be better to just say `between 0 and 306` if you don't want to expand on it too much because without knowing that it also depends on the exponent you start wondering why it's not a single range. For me the only thing that came to mind was that you must have meant "between" very literally and that the edges of the range are not included.

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+- ``uint16(uint8(int8(-1)))`` is ``0xff``.
+
+There are more such changes in relation to payable addresses, literals, enums and contracts. For the full list
+of changes, please see the [documentation](https://docs.soliditylang.org/en/latest/080-breaking-changes.html).
+
+### General Remarks
+
+Since we usually do not backport bugfixes, it is recommended to upgrade all code to be compatible with Solidity v0.8.0.
+
+The [solidity-upgrade tool](https://docs.soliditylang.org/en/latest/using-the-compiler.html#solidity-upgrade) can help you to semi-automatically upgrade your contracts to breaking language changes. While ``solidity-upgrade`` carries out a large part of the work, your contracts will most likely need further manual adjustments.
```
Comment (hrkrshnn) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:218]: `solidity-upgrade` does not do any upgrades for `080`. Maybe we can remove this paragraph?

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+    function f() unchecked { uint x = 7; }
+    ```
+
+    ```solidity
+    function f() pure {
+        uint x;
+        for (uint i = 0; i < 100; i++) unchecked { x += i; }
+    ```
+
+You have to wrap the unchecked blocks in another block to make it work.
```
Comment (cameel) [inline _posts/2020-12-17-solidity-v0.8.0-release-announcement.md:196]: I just noticed that this sentence probably also belongs inside the bullet point.'

```suggestion
    You have to wrap the unchecked blocks in another block to make it work.
```

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+- ``uint16(uint8(int8(-1)))`` is ``0xff``.
+
+There are more such changes in relation to payable addresses, literals, enums and contracts. For the full list
+of changes, please see the [documentation](https://docs.soliditylang.org/en/latest/080-breaking-changes.html).
+
+### General Remarks
+
+Since we usually do not backport bugfixes, it is recommended to upgrade all code to be compatible with Solidity v0.8.0.
+
+The [solidity-upgrade tool](https://docs.soliditylang.org/en/latest/using-the-compiler.html#solidity-upgrade) can help you to semi-automatically upgrade your contracts to breaking language changes. While ``solidity-upgrade`` carries out a large part of the work, your contracts will most likely need further manual adjustments.
```
Comment (hrkrshnn) [inline _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:218]:
```suggestion
```

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+activate the old coder using ``pragma abicoder v1``, or explicitly select v2 using ``pragma abicoder v2``.
+ABI coder v2 is more complex but also performs more checks on the input and supports more types.
+
+Furthermore, internal errors like division by zero, assertion failure and others
+do not use the invalid opcode anymore, but use ``revert`` with a special error message,
+in order to not waste gas in such situations.
+
+Another important change is that we restricted the possibilities of explicit conversions
+to avoid ambiguities. All conversions that were possible before are still possible, but
+you might have to perform two conversions to get there - which does not affect the generated code, though.
```
Comment (hrkrshnn) [inline _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:29]:
```suggestion
you might have to perform two conversions to get there - which does not affect the generated code, when the optimizer is enabled.
```

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+
+These are only some of the more important breaking changes in 0.8.0, please read the changelog below for the full list!
```
Comment (hrkrshnn) [inline _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:31]:
```suggestion
These are only some of the important breaking changes in 0.8.0, please read the changelog below for the full list!
```

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md [hunk 23]
```diff
@@ -0,0 +1,240 @@
...
+
+### Checked Arithmetic
+
+The "Checked Arithmetic" feature of Solidity 0.8.x consists of three sub-features:
+- Revert on assertion failures and similar conditions instead of using the invalid opcode.
+- Actual checked arithmetic.
```
Comment (hrkrshnn) [inline _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:37]:
```suggestion
- Checked arithmetic, i.e., revert on overflows, underflows etc.
```

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md [hunk 24]
```diff
@@ -0,0 +1,240 @@
...
+2. You can only use it inside a block but not as a replacement for a block. The following code snippets, for example, are both invalid:
+
+    ```solidity
+    function f() unchecked { uint x = 7; }
+    ```
+
+    ```solidity
+    function f() pure {
+        uint x;
+        for (uint i = 0; i < 100; i++) unchecked { x += i; }
```
Comment (hrkrshnn) [inline _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:193]:
```suggestion
    ```solidity
    function f() pure {
        uint x;
        for (uint i = 0; i < 100; i++) unchecked { x += i; }
    }
```
Added a missing brace

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md (hunk 1 above)
Comment (franzihei) [reply _posts/2020-12-17-solidity-v0.8.0-release-announcement.md]: I added a sentence wrt the preview release for clarification.

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+published: true
+title: 'Solidity 0.8.0 Release Announcement'
+date: '2020-12-16'
+author: Solidity Team
+category: Releases
+---
+
+[Solidity 0.8.0](https://github.com/ethereum/solidity/releases/tag/v0.8.0) is a breaking release of the Solidity compiler and language. 
+
+Some of the new features of this release have been elaborated in the [0.8.x preview release post](https://blog.soliditylang.org/2020/10/28/solidity-0.8.x-preview/). Please herewith consider the preview release binary superseded and do not use it anymore.
```
Comment (cameel) [inline _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:12]: [Why so formal?](https://memegenerator.net/img/instances/55043458/why-so-formal.jpg)
```suggestion
Some of the new features of this release have been elaborated in the [0.8.x preview release post](https://blog.soliditylang.org/2020/10/28/solidity-0.8.x-preview/). Please consider the preview release binary superseded and do not use it anymore.
```

**Comment by chriseth** [comment]
Will rebase.

File: _posts/2020-12-17-solidity-v0.8.0-release-announcement.md [hunk 26]
```diff
@@ -0,0 +1,240 @@
...
+activate the old coder using ``pragma abicoder v1``, or explicitly select v2 using ``pragma abicoder v2``.
+ABI coder v2 is more complex but also performs more checks on the input and supports more types.
+
+Furthermore, internal errors like division by zero, assertion failure and others
+do not use the invalid opcode anymore, but use ``revert`` with a special error message,
+in order to not waste gas in such situations.
+
+Another important change is that we restricted the possibilities of explicit conversions
+to avoid ambiguities. All conversions that were possible before are still possible, but
+you might have to perform two conversions to get there - which does not affect the generated code, though.
```
[the remark this answers is not in the document]
Comment (chriseth) [reply _posts/2020-12-17-solidity-v0.8.0-release-announcement.md]: Actually no - if you use `uint(int(_a))`, it will generate the same code as for `uint(_a)`.

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md [hunk 27]
```diff
@@ -0,0 +1,277 @@
...
+Note that changes listed below are the **changes between 0.7.6 and 0.8.0**. For changes introduced
+during the 0.7.x series, please see the individual changelogs or release announcements on this blog.
+
+## Full Changelog
+
+### Breaking Changes:
+
+ * Assembler: The artificial ASSIGNIMMUTABLE opcode and the corresponding builtin in the "EVM with object access" dialect of Yul take the base offset of the code to modify as additional argument.
+ * Code Generator: All arithmetic is checked by default. These checks can be disabled using ``unchecked { ... }``.
+ * Code Generator: Cause a panic if a byte array in storage is accessed whose length is encoded incorrectly.
```
Comment (cameel) [inline _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:236]: These indented bullet points might be a problem for some markdown renderers.
```suggestion
* Assembler: The artificial ASSIGNIMMUTABLE opcode and the corresponding builtin in the "EVM with object access" dialect of Yul take the base offset of the code to modify as additional argument.
* Code Generator: All arithmetic is checked by default. These checks can be disabled using ``unchecked { ... }``.
* Code Generator: Cause a panic if a byte array in storage is accessed whose length is encoded incorrectly.
```

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md [hunk 28]
```diff
@@ -0,0 +1,277 @@
...
+You can find a guide on how to update your code [here](https://docs.soliditylang.org/en/latest/080-breaking-changes.html#how-to-update-your-code).
+
+Note that changes listed below are the **changes between 0.7.6 and 0.8.0**. For changes introduced
+during the 0.7.x series, please see the individual changelogs or release announcements on this blog.
+
+## Full Changelog
+
+### Breaking Changes:
+
+* Assembler: The artificial ASSIGNIMMUTABLE opcode and the corresponding builtin in the "EVM with object access" dialect of Yul take the base offset of the code to modify as additional argument.
```
Comment (axic) [inline _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:234]: I would change this in the changelog (and here):
```suggestion
* Yul: The ``assignimmutable` builtin in the "EVM with objects" dialect takes the base offset of the code to modify as an additional argument.
```

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md [hunk 29]
```diff
@@ -0,0 +1,277 @@
...
+* Code Generator: All arithmetic is checked by default. These checks can be disabled using ``unchecked { ... }``.
+* Code Generator: Cause a panic if a byte array in storage is accessed whose length is encoded incorrectly.
+ * Code Generator: Use ``revert`` with error signature ``Panic(uint256)`` and error codes instead of invalid opcode on failing assertions.
```
Comment (axic) [inline _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:237]: There' some indentation change here.

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md [hunk 30]
```diff
@@ -0,0 +1,277 @@
...
+ * Type System: Disallow enums with more than 256 members.
+ * Type System: Disallow explicit conversions from negative literals and literals larger than ``type(uint160).max`` to ``address`` type.
+ * Type System: Introduce ``address(...).code`` to retrieve the code as ``bytes memory``. The size can be obtained via ``address(...).code.length``, but it will currently always include copying the code.
+ * Type System: Explicit conversions between two types are disallowed if it changes more than one of sign, width or kind at the same time.
+ * Type System: Explicit conversions from literals to enums are only allowed if the value fits in the enum.
+ * Type System: Explicit conversions from literals to integer type is as strict as implicit conversions.
+ * Type System: Introduce ``block.chainid`` for retrieving the current chain id.
+ * Type System: Support ``address(...).codehash`` to retrieve the codehash of an account.
+ * Type System: Unary negation can only be used on signed integers, not on unsigned integers.
+ * View Pure Checker: Mark ``chainid`` as view.
```
Comment (axic) [inline _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:260]: This is not sorted.

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md [hunk 31]
```diff
@@ -0,0 +1,277 @@
...
+ * Type System: Explicit conversion to ``address`` type always returns a non-payable ``address`` type. In particular, ``address(u)``, ``address(b)``, ``address(c)`` and ``address(this)`` have the type ``address`` instead of ``address payable`` (Here ``u``, ``b``, and ``c`` are arbitrary variables of type ``uint160``, ``bytes20`` and contract type respectively.)
+ * Type System: The global variables ``tx.origin`` and ``msg.sender`` have type ``address`` instead of ``address payable``.
+ * Yul: Disallow the use of reserved identifiers, such as EVM instructions, even if they are not available in the given dialect / EVM version.
+
+### Language Features:
+ * Super constructors can now be called using the member notation e.g. ``M.C(123)``.
```
Comment (axic) [inline _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:266]: Just wondering, why wasn't this added on 0.7.6 instead? :)

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md [hunk 32]
```diff
@@ -0,0 +1,277 @@
...
+ABI coder v2 is more complex than v1 but also performs additional checks on the input and supports
+a larger set of types than v1.
+
+Furthermore, internal errors like division by zero, assertion failure and others
+do not use the invalid opcode anymore, but use ``revert`` with a special error message,
+in order to not waste gas in such situations.
+
+Another important change is that we restricted the possibilities of explicit conversions
+to avoid ambiguities. All conversions that were possible before are still possible, but
+you might have to perform two conversions to get there - which does not affect the generated code, when the optimizer is enabled.
```
Comment (hrkrshnn) [inline _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:35]:
```suggestion
you might have to perform two conversions to get there - which does not affect the generated code, though.
```
@chriseth should we revert this change?

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md (hunk 29 above)
Comment (cameel) [reply _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:237]: Looks like my suggestion was applied but it wasn't complete - I marked only the first 3 lines because the list is long.

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md (hunk 28 above)
Comment (chriseth) [reply _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:234]: Oh come on, this is the changelog...

Comment (chriseth) [reply _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:234]: I won't change it in the other two places.

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md (hunk 31 above)
Comment (chriseth) [reply _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:266]: It's a breaking change

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md (hunk 32 above)
Comment (chriseth) [reply _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:35]: ah right!

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md (hunk 30 above)
Comment (chriseth) [reply _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:260]: *sigh*.

File: _posts/2020-12-16-solidity-v0.8.0-release-announcement.md (hunk 29 above)
Comment (chriseth) [reply _posts/2020-12-16-solidity-v0.8.0-release-announcement.md:237]: Will relpace the whole changelog.
